File: advent4.py
f = open('advent.txt','r')
text = f.read()
text = text.split('\n')
text = [line.split(',') for line in text]

counter = 0

def compare(a,b,c):
    return a <= b and b <= c

def overlap(sched):
    if compare(sched[0][0],sched[1][0],sched[0][1]) or compare(sched[0][0],sched[1][1],sched[0][1]):
        return True
    elif compare(sched[1][0],sched[0][0],sched[1][1]) or compare(sched[1][0],sched[0][1],sched[1][1]):
        return True
    else:
        return False

for line in text:
    sched = [s.split('-') for s in line]
    sched = [[int(x) for x in y] for y in sched]
    
    if overlap(sched):
        counter += 1

print(counter)

f.close()

File: advent4.rs
extern crate regex;
use regex::Regex;
use std::fs;

fn overlap(x : Vec<&str>) -> u16 {
    let x: Vec<u16> = x.into_iter().map(|y| y.parse::<u16>().unwrap()).collect();
    if x[0] <= x[2] && x[1] >= x[3]{
        return 1;
    }
    else if x[0] >= x[2] && x[1] <= x[3] {
        return 1;
    }
    else {
        return 0;
    }
}
fn main() {
    let file_path: &str = "./advent.txt";
    let schedules= fs::read_to_string(file_path).expect("Should have been able to read the file");
    let schedules: Vec<&str> = schedules.split("\r\n").collect();
    let re = Regex::new(r"\D+").expect("Invaled regex");
    let schedules: Vec<Vec<&str>> = schedules
        .into_iter()
        .map(|x| re
            .split(x)
            .collect::<Vec<&str>>()
        )
        .collect();
    let schedules: Vec<u16> = schedules
        .into_iter()
        .map(|x| overlap(x))
        .collect();
    let answer: u16 = schedules.into_iter().sum();

    println!("{answer:#?}");
}